hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/metadata/SparkHoodieMetadataBulkInsertPartitioner.java

```diff
@@ -0,0 +1,111 @@
+    this.numPartitions = numPartitions;
+  }
+
+  private class FileGroupPartitioner extends Partitioner {
+
+    @Override
+    public int getPartition(Object key) {
+      return ((Tuple2<Integer, String>)key)._1;
+    }
+
+    @Override
+    public int numPartitions() {
+      return numPartitions;
+    }
+  }
+
+  // FileIDs for the various partitions
+  private List<String> fileIDPfxs;
+
+  /**
+   * Partition the records by their location. The number of partitions is determined by the number of MDT fileGroups being udpated rather than the
+   * specific value of outputSparkPartitions.
+   */
+  @Override
+  public JavaRDD<HoodieRecord> repartitionRecords(JavaRDD<HoodieRecord> records, int outputSparkPartitions) {
+    Comparator<Tuple2<Integer, String>> keyComparator =
+            (Comparator<Tuple2<Integer, String>> & Serializable)(t1, t2) -> t1._2.compareTo(t2._2);
+
+    // Partition the records by their file group
+    JavaRDD<HoodieRecord> partitionedRDD = records
+            // key by <file group index, record key>. The file group index is used to partition and the record key is used to sort within the partition.
+            .keyBy(r -> {
+              int fileGroupIndex = HoodieTableMetadataUtil.getFileGroupIndexFromFileId(r.getCurrentLocation().getFileId());
+              return new Tuple2<>(fileGroupIndex, r.getRecordKey());
+            })
+            .repartitionAndSortWithinPartitions(new FileGroupPartitioner(), keyComparator)
+            .map(t -> t._2);
+
+    fileIDPfxs = partitionedRDD.mapPartitions(recordItr -> {
+      // Due to partitioning, all record in the partition should have same fileID. So we only can get the fileID prefix from the first record.
+      List<String> fileIds = new ArrayList<>(1);
+      if (recordItr.hasNext()) {
+        HoodieRecord record = recordItr.next();
+        final String fileID = HoodieTableMetadataUtil.getFileGroupPrefix(record.getCurrentLocation().getFileId());
+        fileIds.add(fileID);
+      } else {
+        // FileGroupPartitioner returns a fixed number of partition as part of numPartitions(). In the special case that recordsRDD has fewer
+        // records than fileGroupCount, some of these partitions (corresponding to fileGroups) will not have any data.
+        // But we still need to return a fileID for use within {@code BulkInsertMapFunction}
+        fileIds.add("");
+      }
+      return fileIds.iterator();
+    }, true).collect();
+    ValidationUtils.checkArgument(partitionedRDD.getNumPartitions() == fileIDPfxs.size(),
+            String.format("Generated fileIDPfxs (%d) are lesser in size than the partitions %d", fileIDPfxs.size(), partitionedRDD.getNumPartitions()));
+
+    return partitionedRDD;
+  }
+
+  @Override
+  public boolean arePartitionRecordsSorted() {
+    return true;
+  }
+
+  @Override
+  public String getFileIdPfx(int partitionId) {
+    return fileIDPfxs.get(partitionId);
```

Comment:
This may not be right when one Hfile in one file group is too large to write with, there is a upper threshold for each base file handle, `HoodieAvroHFileWrite#canWrite` may return false, then the write handle factory would suffix the file group id with another auto inc number like `-1` which is not correct any more.

Reply:
yes. this is known. we have set file size as 1GB. so, users have to set the right config for num of file groups.

Reply:
No one can have awareness the file group number is pertinent with the correctness. It is a bug, not a usability issue.

Reply:
The idea behind sharding in MDT is that you can create more shards rather than splitting a shard into two files. Spliiting will cause one large and one small size file which is not optimal.
 For optimal performance, all shards should be of similar size which we achieve by hash partitioning the keys.

Reply:
This PR also adds support for automatic estimation of the shard counts for each partition, that can be enhanced.